services/api/src/migrations/lagoon/migrations/20231130000000_group_projects.ts

```diff
@@ -0,0 +1,45 @@
+import * as R from 'ramda';
+import { logger } from '@lagoon/commons/dist/logs/local-logger';
+import { getKeycloakAdminClient } from '../../../clients/keycloak-admin';
+import { sqlClientPool } from '../../../clients/sqlClient';
+import { esClient } from '../../../clients/esClient';
+import redisClient from '../../../clients/redisClient';
+import { Group } from '../../../models/group';
+import { Helpers } from '../../../resources/group/helpers';
+
+export const up = async (migrate) => {
+  const keycloakAdminClient = await getKeycloakAdminClient();
+
+  const GroupModel = Group({
+    sqlClientPool,
+    keycloakAdminClient,
+    esClient,
+    redisClient
+  });
+
+  // load all groups from keycloak
+  const allGroups = await GroupModel.loadAllGroups();
+  // transform them into lagoon expected format
+  const keycloakGroups = await GroupModel.transformKeycloakGroups(allGroups);
+  for (const kg of keycloakGroups) {
+    // extract the project ids from the group
+    const groupProjects = await GroupModel.getProjectsFromGroupAndSubgroups(kg);
```

Comment:
I don't think `getProjectsFromGroupAndSubgroups` returns the desired result here? That is designed to calculate the projects a user has access to, taking into account group hierarchy. It will return projects that are not tied directly to the group being iterated (`kg`). But this migration is to copy 1-1 the data in keycloak, which should be tying projects only to the groups they are explicitly added to.

I think instead you would [flatten](https://github.com/uselagoon/lagoon/blob/main/services/api/src/models/group.ts#L198-L206) the `loadAllGroups` result, and [filter](https://github.com/uselagoon/lagoon/blob/main/services/api/src/models/group.ts#L147) out role groups. The remaining should be regular groups that you can iterate over and [getProjectIdsFromGroup](https://github.com/uselagoon/lagoon/blob/main/services/api/src/models/group.ts#L94-L101).